python/dgl/dataloading/negative_sampler.py

```diff
@@ -0,0 +1,68 @@
+"""Negative samplers"""
+from collections.abc import Mapping
+import numpy as np
+from .. import backend as F
+
+class _BaseNegativeSampler(object):
+    def _generate(self, g, eids, canonical_etype):
+        raise NotImplementedError
+
+    def __call__(self, g, eids):
+        """Returns negative examples.
+
+        Parameters
+        ----------
+        g : DGLHeteroGraph
+            The graph.
+        eids : Tensor or dict[etype, Tensor]
+            The sampled edges in the minibatch.
+
+        Returns
+        -------
+        tuple[Tensor, Tensor] or dict[etype, tuple[Tensor, Tensor]]
+            The returned source-destination pairs as negative examples.
+        """
+        if isinstance(eids, Mapping):
+            eids = {g.to_canonical_etype(k): v for k, v in eids.items()}
+            neg_pair = {k: self._generate(g, v, k) for k, v in eids.items()}
+        else:
+            assert len(g.etypes) == 1, \
+                'please specify a dict of etypes and ids for graphs with multiple edge types'
+            neg_pair = self._generate(g, eids, g.canonical_etypes[0])
+
+        return neg_pair
+
+class Uniform(_BaseNegativeSampler):
+    """Negative sampler that randomly chooses negative destination nodes
+    for each source node according to a uniform distribution.
+
+    For each edge with type `(utype, etype, vtype)`, ``k`` pairs of nodes
+    with node type ``utype`` and ``vtype`` will be returned.  The source nodes
+    will always be the source node of the edge, while the destination nodes
```

Comment:
What about negative source nodes?

Reply:
The most common use case is to fix the source node and sample the negative destination nodes.  If you really want negative source nodes you could write your own negative sampler as it's relatively easy (see `graphsage/train_sampling_unsupervised.py`)